sdk/identity/azure-identity/src/client_certificate_credential.cpp

```diff
@@ -50,6 +60,204 @@ template <typename T> std::vector<uint8_t> ToUInt8Vector(T const& in)
+using PrivateKey = decltype(std::declval<UniquePrivateKey>().get());
+
+#ifdef AZ_PLATFORM_WINDOWS
+enum PrivateKeyType
+{
+  Rsa,
+  Ecdsa,
+  Pkcs
+};
+
+std::vector<uint8_t> PemToBinary(LPCSTR str, DWORD count)
+{
+  DWORD size = 0;
+  THROW_IF_WIN32_BOOL_FALSE(CryptStringToBinaryA(
+      str, count, CRYPT_STRING_BASE64HEADER, nullptr, &size, nullptr, nullptr));
+  std::vector<uint8_t> buffer(size);
+  THROW_IF_WIN32_BOOL_FALSE(CryptStringToBinaryA(
+      str, count, CRYPT_STRING_BASE64HEADER, buffer.data(), &size, nullptr, nullptr));
+  return buffer;
+}
+
+CertificateThumbprint GetThumbprint(PCCERT_CONTEXT cert)
+{
+  DWORD size = 0;
+  THROW_IF_WIN32_BOOL_FALSE(
+      CertGetCertificateContextProperty(cert, CERT_SHA1_HASH_PROP_ID, nullptr, &size));
+  std::vector<unsigned char> thumbprint(size);
+  THROW_IF_WIN32_BOOL_FALSE(
+      CertGetCertificateContextProperty(cert, CERT_SHA1_HASH_PROP_ID, thumbprint.data(), &size));
+  return thumbprint;
+}
+
+wil::unique_cert_context ImportPemCertificate(std::string const& pem)
+{
+  auto headerStart = pem.find("-----BEGIN CERTIFICATE-----");
+  if (headerStart == std::string::npos)
+  {
+    throw AuthenticationException("PEM file does not contain certificate.");
+  }
+  auto certBuffer
+      = PemToBinary(pem.c_str() + headerStart, static_cast<DWORD>(pem.size() - headerStart));
+  auto cert = CertCreateCertificateContext(
+      X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+      certBuffer.data(),
+      static_cast<DWORD>(certBuffer.size()));
+  THROW_LAST_ERROR_IF_NULL(cert);
+  return wil::unique_cert_context{cert};
+}
+
+size_t FindPemPrivateKeyHeader(std::string const& pem, PrivateKeyType& keyType)
+{
+  auto headerStart = pem.find("-----BEGIN RSA PRIVATE KEY-----");
+  if (headerStart != std::string::npos)
+  {
+    keyType = PrivateKeyType::Rsa;
+    return headerStart;
+  }
+  headerStart = pem.find("-----BEGIN EC PRIVATE KEY-----");
+  if (headerStart != std::string::npos)
+  {
+    keyType = PrivateKeyType::Ecdsa;
+    return headerStart;
+  }
+  keyType = PrivateKeyType::Pkcs;
+  return pem.find("-----BEGIN PRIVATE KEY-----");
+}
+
+wil::unique_bcrypt_algorithm OpenAlgorithm(LPCWSTR algId)
+{
+  wil::unique_bcrypt_algorithm alg;
+  THROW_IF_NTSTATUS_FAILED(BCryptOpenAlgorithmProvider(wil::out_param(alg), algId, nullptr, 0));
+  return alg;
+}
+
+UniquePrivateKey ImportRsaPrivateKey(const BYTE* data, DWORD size)
+{
+  DWORD keySize = 0;
+  wil::unique_hlocal_ptr<BCRYPT_RSAKEY_BLOB> rsaKeyBlob;
+  THROW_IF_WIN32_BOOL_FALSE(CryptDecodeObjectEx(
+      X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+      CNG_RSA_PRIVATE_KEY_BLOB,
+      data,
+      size,
+      CRYPT_DECODE_ALLOC_FLAG,
+      nullptr,
+      wil::out_param(rsaKeyBlob),
+      &keySize));
+  BCRYPT_KEY_HANDLE key;
+  auto alg = OpenAlgorithm(BCRYPT_RSA_ALGORITHM);
+  THROW_IF_NTSTATUS_FAILED(BCryptImportKeyPair(
+      alg.get(),
+      nullptr,
+      BCRYPT_RSAPRIVATE_BLOB,
+      &key,
```

Comment:
You may want to use UniquePrivateKey.get_address_of to avoid potential leaks.

Reply:
`UniquePrivateKey` does not have `get_address_of` or any related member. I don't think the code is prone to leaks right now; right after we check that `BCryptImportKeyPair` succeeded, we create the `UniquePrivateKey`.

Reply:
Ah - the wil variants don't, but the __detail::unique_handle does not.